Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+      2. based on both availability of free pages and type of content asosciated, the access might cause the process
+         to sleep; exaples:
+         1. no free pages are available (unusual but possible) and the kernel will have to try to obtain one, in one of
+            the ways described above
+         2. the content needs to be loaded from a file, and the operation is blocking, because the retrieval process is much slower.
+      3. the only exception, where pages are specifically allocated before the process is started, is the beginning of the stack; here process arguments and context are stored. 
+      Either way, it is difficult to know if/when a process is not going to generate any more faults, and it is very much not deterministic.
+5. creations of threads for an existing process reuses the process memory map almost entirely, however an additional stack is
+   allocated for each new thread, picking an address range between the live ends of heap and primary stack.
+   The sizes of these additional stacks are defined at creation time, but allocation of physical pages happens on-demand.
+6. as also decribed in the previous points, the kernel uses various optimisations for dealing with processes on-demand mapping:
+   1. a physical page is allocated and mapped to a process only when the process accesses it, otherwise it might not be present:
+      1. file backed pages are allocated/mapped only when read/written to
+      2. anonymous pages are allocated only when written to
+   2. zero page: when a page is known to be empty, it is not reserved and mapped; instead,
+      the kernel has one specific (dummy) memory page that is mapped tothe process as read-only
+   3. shared libraries:
+      when the same library pages are mapped by multiple processes, the library physical pages
+      are mapped as read-only into each process address space
+   4. copy-on-write:
+      1. when a library has own data, this is initially mapped as read-only and shared;
+         only when written to, then a separate physical page is reserved for each writing process (threads sahre the same)
+      2. same happens for data pages that were treated as zero-page, but then are written to.
+   5. folios (see also kernel section): data structures that try to better abstract compound pages and *might* also be used
+      to represent optimised contiguous pages on ARM64 (instead of mapping 16 entries, it is possible
+      to map a 16-pages chunk of physically contiguous memory, that is also aligned to a 16 pages
+      boundary, reducing TLB use).
+      1. A folio acts as intermediary between vma and lower level memory management
+      2. it might pre-allocate/map more pages than explicitly requested
+7. read-ahead: when asked to fetch data from disk, the kernel might attempt to optimise the operation,
+   reading more pages than requested, under the assumption that more requests might be coming soon
+ 
+
+#### **Safety-Oriented consideration**
```

Comment:
Same comment as above. We should to separate the premises from these considerations, and explicitly state the premises each conclusion builds on. Take for example the 1 argument below. Uncertainty in the timing might be a problem, but might not, depending on the requirements or on the size of the uncertainty.

Reply:
Agreed to reconsider later the overall posture of the document.
My motivation is to try to keep together statements which are kind of obvious with others that might be less obvious, but derivable.
We could consider - in general - to introduce also the notion of  which aspects are likely to be unavoidable, and which ones might be relevant only for some use case scenarios.